File: haskell/src/Solutions/Euler510.hs
module Solutions.Euler510 where

funS n = sum [(raBase + rbBase + rcBase) * maxK * (maxK + 1) `div` 2 |
    beta <- [1..floor $ fromIntegral n ** (1 / 4)],
    alpha <- [1..beta],
    gcd alpha beta == 1,
    let alpha2 = alpha^2,
    let beta2 = beta^2,
    let den = (alpha + beta)^2,
    let maxK = n `div` (beta2 * den),
    let raBase = den * alpha2,
    let rbBase = den * beta2,
    let rcBase = alpha2 * beta2]

euler510 :: IO String
euler510 = return $ show $ funS $ 10^9

File: rust/src/solutions/euler510.rs
fn s(n: u64) -> u64 {
    let mut ans = 0;
    for beta in 1u64..=(n as f64).powf(0.25).floor() as u64 {
        for alpha in (1..=beta).filter(|&alpha| num::integer::gcd(alpha, beta) == 1) {
            let alpha2 = alpha * alpha;
            let beta2 = beta * beta;
            let den = (alpha + beta) * (alpha + beta);
            let max_k = n / (beta2 * den);
            let ra_base = den * alpha2;
            let rb_base = den * beta2;
            let rc_base = alpha2 * beta2;
            ans += (ra_base + rb_base + rc_base) * max_k * (max_k + 1) / 2;
        }
    }
    ans
}

pub fn euler510() -> String {
    s(1_000_000_000).to_string()
}
